pkg/sensors/tracing/genericlsm.go

```diff
@@ -0,0 +1,465 @@
+	}
+
+	tagsField, err := getPolicyTags(f.Tags)
+	if err != nil {
+		return errFn(fmt.Errorf("Error: '%v'", err))
+	}
+
+	// Parse Arguments
+	for j, a := range f.Args {
+		argType := gt.GenericTypeFromString(a.Type)
+		if argType == gt.GenericInvalidType {
+			return errFn(fmt.Errorf("Arg(%d) type '%s' unsupported", j, a.Type))
+		}
+		if a.MaxData {
+			if argType != gt.GenericCharBuffer {
+				logger.GetLogger().Warnf("maxData flag is ignored (supported for char_buf type)")
+			}
+			if !kernels.EnableLargeProgs() {
+				logger.GetLogger().Warnf("maxData flag is ignored (supported from large programs)")
+			}
+		}
+		argMValue, err := getMetaValue(&a)
+		if err != nil {
+			return errFn(err)
+		}
+		if a.Index > 4 {
+			return errFn(fmt.Errorf("Error add arg: ArgType %s Index %d out of bounds",
+				a.Type, int(a.Index)))
+		}
+		config.Arg[a.Index] = int32(argType)
+		config.ArgM[a.Index] = uint32(argMValue)
+
+		argsBTFSet[a.Index] = true
+		argP := argPrinter{index: j, ty: argType, maxData: a.MaxData, label: a.Label}
+		argSigPrinters = append(argSigPrinters, argP)
+	}
+
+	config.ArgReturn = int32(0)
+	config.ArgReturnCopy = int32(0)
+
+	// Mark remaining arguments as 'nops' the kernel side will skip
+	// copying 'nop' args.
+	for j, a := range argsBTFSet {
+		if !a {
+			if j != api.ReturnArgIndex {
+				config.Arg[j] = gt.GenericNopType
+				config.ArgM[j] = 0
+			}
+		}
+	}
+
+	config.Syscall = 0
+
+	// create a new entry on the table, and pass its id to BPF-side
+	// so that we can do the matching at event-generation time
+	lsmEntry := genericLsm{
+		config:      config,
+		argPrinters: argSigPrinters,
+		hook:        f.Hook,
+		tableId:     idtable.UninitializedEntryID,
+		policyName:  in.policyName,
+		message:     msgField,
+		tags:        tagsField,
+	}
+
+	// Parse Filters into kernel filter logic
+	lsmEntry.selectors, err = selectors.InitKernelSelectorState(f.Selectors, f.Args, nil, nil, in.selMaps)
+	if err != nil {
+		return errFn(err)
+	}
+
+	genericLsmTable.AddEntry(&lsmEntry)
+	config.FuncId = uint32(lsmEntry.tableId.ID)
+
+	lsmEntry.pinPathPrefix = sensors.PathJoin(in.sensorPath, fmt.Sprintf("glsm-%d", lsmEntry.tableId.ID))
+
+	logger.GetLogger().
+		WithField("hook", lsmEntry.hook).
+		Infof("Added lsm Hook")
+
+	return lsmEntry.tableId, nil
+}
+
+func createGenericLsmSensor(
+	spec *v1alpha1.TracingPolicySpec,
+	name string,
+	policyID policyfilter.PolicyID,
+	policyName string,
+) (*sensors.Sensor, error) {
+	var progs []*program.Program
+	var maps []*program.Map
+	var ids []idtable.EntryID
+	var selMaps *selectors.KernelSelectorMaps
+	var err error
+
+	if !kernels.MinKernelVersion("5.7") {
+		return nil, fmt.Errorf("to load LSM sensor kernel version must be greater or equal 5.7")
+	}
+
+	if !bpf.HasLSMPrograms() {
+		return nil, fmt.Errorf("enable LSM BPF by modifying the GRUB configuration /etc/default/grub with GRUB_CMDLINE_LINUX=\"lsm=bpf\"")
+	}
+
+	lsmHooks := spec.LsmHooks
```

Comment:
At this point, I think we should check the actions and reject policies with actions that are not properly supported in the LSM sensor.

One example is Post actions with stack trace, since in ` handleGenericLsm` we do not seem to do what is done in kprobes: https://github.com/cilium/tetragon/blob/e44f7dfc0e8ee90ddcef1d915111d420917b9ae0/pkg/sensors/tracing/generickprobe.go#L1217

Another example is the GetURL, DnsLookup actions: https://github.com/cilium/tetragon/blob/e44f7dfc0e8ee90ddcef1d915111d420917b9ae0/pkg/sensors/tracing/generickprobe.go#L1177

Unfortunately, there is a lot of code duplication among the kprobe/tracepoint/uprobe and now the LSM sensor. It would be nice if we could reduce it, but that's a different issue alltogether.

Reply:
Ok, I'll carefully look what actions aren't available from LSM hooks and check for them.

Yes, I noticed that there are some code duplicates among kprobes/uprobes/tracepoints/lsm. Also data structures from kprobe sensor are used in another types of programs.  I agree, it is a good idea to do some refactoring in that case in future.